packages/rgbpp/src/rgbpp/xudt/btc-transfer-all.ts

```diff
@@ -0,0 +1,219 @@
+import { buildXudtTypeScriptHex, encodeCellId, unpackRgbppLockArgs } from '../utils/ckb';
+import { RgbppTransferAllTxGroup, RgbppTransferAllTxsParams, RgbppTransferAllTxsResult } from '../types/xudt';
+import { AssetSummarizer } from '../summary/asset-summarizer';
+import { RgbppError, RgbppErrorCodes } from '../error';
+
+/**
+ * Build BTC/CKB transaction group(s) to transfer all amount of an RGB++ XUDT asset from the specified BTC addresses to the target BTC address.
+ * Due to the input size limitation, each CKB transaction can contain up to 40 RGB++ cells.
+ * Therefore, the generated result may contain more than one transaction group.
+ */
+export async function buildRgbppTransferAllTxs(params: RgbppTransferAllTxsParams): Promise<RgbppTransferAllTxsResult> {
+  // Prepare base props
+  const maxRgbppCellsPerCkbTx = RGBPP_TX_INPUTS_MAX_LENGTH;
+  const isMainnet = params.isMainnet;
+  const btcSource = params.btc.dataSource;
+  const btcService = btcSource.service;
+  const ckbCollector = params.ckb.collector;
+  const xudtTypeHex = buildXudtTypeScriptHex(params.ckb.xudtTypeArgs, isMainnet);
+
+  // Get L2 Cells own by the assetAccounts,
+  // and build L1 UTXO IDs (`${txid}:${vout}`) from each cell.cellOutput.lock.args
+  const accountCells = await Promise.all(
+    params.btc.assetAddresses.map((address) => {
+      return limitPromiseBatchSize(() =>
+        btcService.getRgbppAssetsByBtcAddress(address, {
+          type_script: xudtTypeHex,
+        }),
+      );
+    }),
+  );
+  const utxoIds = new Set<string>(
+    accountCells.flat().map((rgbppCell) => {
+      const lockArgs = unpackRgbppLockArgs(rgbppCell.cellOutput.lock.args);
+      return encodeUtxoId(lockArgs.btcTxid, lockArgs.outIndex);
+    }),
+  );
+
+  // Get all L1 UTXOs from the asset accounts
+  const accountUtxos = await Promise.all(
+    params.btc.assetAddresses.map((address) => {
+      return limitPromiseBatchSize(() => btcSource.getUtxos(address));
+    }),
+  );
+
+  // Storage variables for cache querying
+  const invalidUtxoIds = new Set<string>(); // Set<UtxoId>
+  const cellsMap = new Map<string, Cell[]>(); // Map<UtxoId, Cell[]>
+  const utxosMap = new Map<string, Utxo>(); // Map<CellId, Utxo>
+  const cellMap = new Map<string, Cell>(); // Map<CellId, Cell>
+  const utxoMap = accountUtxos.flat().reduce(
+    // Map<UtxoId, Utxo>
+    (map, utxo) => {
+      const utxoId = encodeUtxoId(utxo.txid, utxo.vout);
+      if (utxoIds.has(utxoId)) {
+        map.set(utxoId, utxo);
+      }
+      return map;
+    },
+    new Map<string, Utxo>(),
+  );
+
+  // Get each L1 UTXO's corresponding L2 Cells
+  // If a UTXO has more L2 Cells than maxRgbppCellsPerCkbTx, the UTXO/Cells are invalid.
+  // Invalid L1 UTXO and its corresponding L2 Cells are excluded from the transfer process.
+  const rgbppGroups: { id: string; cells: number }[] = [];
+  await Promise.all(
+    [...utxoIds].map(async (utxoId) => {
```

Comment:
Is using `utxoIds` instead of `[...utxoIds]` more simple?

Reply:
You're right, I'm not sure why I wrote a clone here, I'll check and fix it.

Reply:
I double-checked the code, turns out this part is expected because `utxoIds` is a `Set<string>` object and does not have a `.map()` method, which is why I used the `[...utxoIds]` syntax.